File: javascript/binarySearch.js


var search = function(nums, target) {
    var len = nums.length;
    var smaller = 0;
    var bigger = len - 1;
    //console.log(bigger);
    if (len === 1) {
        return(nums[0] === target ? 0 : -1);
    }
    while(smaller <= bigger) {
        var mid = Math.floor((smaller + bigger) / 2);
        if (smaller === bigger - 1){
            if(nums[smaller] === target) {
                return smaller;
            } else if (nums[bigger] === target) {
                return bigger;
            } else {
                break;
            }
        }
        if (nums[mid] > target) {
            bigger = mid;
            continue;
        } else if (nums[mid] < target) {
            smaller = mid;
            continue;
        } else {
            return mid;
        }
    }
    return -1;
 };

File: C/binarySearch.c
int search(int *nums, int numsSize, int target)
{
    int small = 0;
    int big = numsSize - 1;

    if (numsSize == 1)
    {
        if (nums[0] == target)
        {
            return 0;
        }
        else
        {
            return -1;
        }
    }
    while (small <= big)
    {
        int mid = (small + big) / 2;
        if (small == big - 1)
        {
            if (nums[mid] == target)
            {
                return small;
            }
            else if (nums[big] == target)
            {
                return big;
            }
            else
            {
                break;
            }
        }
        if (nums[mid] > target)
        {
            big = mid;
        }
        else if (nums[mid] < target)
        {
            small = mid;
        }
        else
        {
            return mid;
        }
    }
    return -1;
}